standards/practices/testing.md

```diff
@@ -0,0 +1,194 @@
+2. One test covers two different actions.
+3. The test name is difficult to read in camelCase.
+4. The test mixes all stages together.
+5. The code covered is so simple it's not worth testing (little more than third-party code hidden behind a simple ternary).
+
+**Good**
+```php
+<?php
+
+namespace Tests\Feature;
+
+use App\ClientUser;
+use App\User;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Tests\TestCase;
+
+class AuthenticationTest extends TestCase
+{
+    use RefreshDatabase;
+
+    /** @test */
+    public function admins_can_log_in()
+    {
+        // Arrange
+        // Create an admin.
+        $user = factory(User::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+
+    /** @test */
+    public function client_users_can_log_in()
+    {
+        // Arrange
+        // Create a client user.
+        $user = factory(ClientUser::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+}
+```
+
+*In most frameworks, it's not worth testing authentication, because the framework handles pretty much everything. However in this example, the authentication is customized enough that we need to avoid regressions.*
+
+These two tests improve on the previous example in several ways:
+
+1. We converted from a unit test to an integration test (called a feature test in Laravel). The feature depends on several systems more simply tested with a couple short integration tests than with several unit tests that can't reliably test the whole feature.
+2. Each test covers only one action.
+3. We formatted the test names so people can easily read it and understand what it tests.
+4. Each test is split into the standard arrange, act, and assert steps with helpful explanations.
+5. The tests cover complex functions likely affected by regressions.
+
+
+## Explanation
+Tests help us improve the quality of our software. However, tests carry maintenance costs, so **our tests should add value to our software**. In pursuit of this value:
+
+1. **Avoid overly-specific coverage goals.**  
+  E.g. 80% code coverage. Goals like this incentivize writing more tests than required to obtain confidence in the software.
+2. **Don't test what you don't own.**  
+  Testing frameworks and other third-party software indicates you don't trust the package you decided to rely on.
```

Comment:
I feel like we can have a 3rd point here.  We shouldn't try to test overly simple code.  In my original write up I said "don't test 1+1=2".  You put something similar: "The code covered is so simple it's not worth testing (little more than third-party code hidden behind a simple ternary)."

Reply:
Yeah, I want to make that same point w/o actually saying "1 + 1 = 2" b/c it's easy for people to argue past each other with an example that (hopefully) abstract.

What do you think of this?

3. **Avoid testing basic operations.**  
  Testing that PHP itself functions properly wastes effort and dilutes test suite value.